crates/lib/src/qpu/experimental/randomized_measurements.rs

```diff
@@ -0,0 +1,628 @@
+        Ok(Self {
+            leading_delay,
+            unitary_set,
+            unitary_count_as_u8,
+            qubit_randomizations,
+        })
+    }
+}
+
+impl RandomizedMeasurements {
+    /// Append the randomized measurements to a Quil program. The provided program
+    /// should not contain any preexisting measurements.
+    pub fn append_to_program(&self, target_program: Program) -> Result<Program> {
+        let mut program = target_program.clone_without_body_instructions();
+
+        program.add_instruction(Instruction::Declaration(Declaration {
+            name: RANDOMIZED_MEASUREMENT_SOURCE.to_string(),
+            size: Vector::new(
+                ScalarType::Real,
+                (self.unitary_set.parameters_per_unitary() * self.unitary_set.unitary_count())
+                    as u64,
+            ),
+            sharing: None,
+        }));
+
+        for qubit_randomization in &self.qubit_randomizations {
+            program.add_instruction(Instruction::Declaration(
+                qubit_randomization.destination_declaration.clone(),
+            ));
+            program.add_instruction(Instruction::Declaration(
+                qubit_randomization.seed_declaration.clone(),
+            ));
+        }
+
+        // prepend delay
+        program.add_instruction(Instruction::Delay(Delay::new(
+            self.leading_delay.clone(),
+            Vec::new(),
+            self.qubit_randomizations
+                .iter()
+                .map(|randomization| randomization.measurement.qubit.clone())
+                .collect(),
+        )));
+
+        // declare "choose_random_real_sub_regions" as an external function
+        program.add_instruction(Instruction::Pragma(
+            super::random::ChooseRandomRealSubRegions::pragma_extern()?,
+        ));
+
+        let parameters_per_unitary = self
+            .unitary_set
+            .parameters_per_unitary()
+            .to_f64()
+            .ok_or_else(|| {
+                Error::ParametersPerUnitaryF64Conversion(self.unitary_set.parameters_per_unitary())
+            })?;
+        // Before the pulse program begins, set the randomized unitary for each qubit.
+        let calls: Vec<_> = self
+            .qubit_randomizations
+            .iter()
+            .cloned()
+            .map(|qubit_randomization| qubit_randomization.into_instruction(parameters_per_unitary))
+            .collect::<Result<Vec<Instruction>>>()?;
+        program.add_instructions(calls);
+
+        // Include the program body that was passed in.
+        program.add_instructions(target_program.into_body_instructions());
+
+        // Play the random unitaries on each qubit.
+        program.add_instructions(
+            self.unitary_set
+                .to_instructions(self.qubit_randomizations.as_slice()),
+        );
+
+        program.add_instruction(Instruction::Fence(Fence { qubits: Vec::new() }));
```

Comment:
Because we want to play the 1Q rotations and measurements as two sequential and isolated cycles regardless as to what else is going on in the program.